compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -424,5 +430,218 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
+      workgroupCount.push_back(workload[tileSize.index()]);
+      continue;
+    }
+    AffineExpr s0, s1;
+    bindSymbols(rewriter.getContext(), s0, s1);
+    auto m = AffineMap::get(0, 2, s0.ceilDiv(s1));
+    Value tileSizeVal =
+        getValueOrCreateConstantIndexOp(rewriter, loc, tileSize.value());
+    Value count = rewriter.create<AffineApplyOp>(
+        loc, m, ValueRange{workload[tileSize.index()], tileSizeVal});
+    workgroupCount.push_back(count);
+  }
+  workgroupCount = llvm::to_vector(llvm::reverse(workgroupCount));
+  workgroupCount.resize(3, rewriter.create<arith::ConstantIndexOp>(loc, 1));
+  rewriter.replaceOp(workgroupCountOp, workgroupCount);
+  return success();
+}
+
+SmallVector<OpFoldResult>
+transform_dialect::TileToWorkgroupsOp::getMixedNumThreads() {
+  return getMixedSizes(getStaticNumThreads(), getNumThreads());
+}
+
+SmallVector<OpFoldResult>
+transform_dialect::TileToWorkgroupsOp::getMixedTileSizes() {
+  return getMixedSizes(getStaticTileSizes(), getTileSizes());
+}
+
+LogicalResult transform_dialect::TileToWorkgroupsOp::verify() {
+  if (getMixedNumThreads().empty() == getMixedTileSizes().empty())
+    return emitOpError("either num_threads or tile_sizes must be specified");
+  return success();
+}
+
+void transform_dialect::TileToWorkgroupsOp::getEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  transform::consumesHandle(getTarget(), effects);
+  transform::consumesHandle(getFunc(), effects);
+  transform::onlyReadsHandle(getTileSizes(), effects);
+  transform::onlyReadsHandle(getNumThreads(), effects);
+  transform::producesHandle(getResults(), effects);
+}
+
+DiagnosedSilenceableFailure transform_dialect::TileToWorkgroupsOp::apply(
+    transform::TransformResults &transformResults,
+    transform::TransformState &state) {
+  ArrayRef<Operation *> funcOps = state.getPayloadOps(getFunc());
+  assert(funcOps.size() == 1 && "expected single func op in payload");
+  FailureOr<IREE::HAL::ExecutableExportOp> exportOp =
+      getEntryPoint(cast<func::FuncOp>(funcOps[0]));
+  if (failed(exportOp)) {
+    state.getTopLevel()->emitOpError("couldn't find export op for func");
+    return DiagnosedSilenceableFailure(reportUnknownTransformError(funcOps[0]));
+  }
+
+  SmallVector<OpFoldResult> mixedTileSizes = getMixedTileSizes();
+  if (mixedTileSizes.empty()) {
+    exportOp.value()->emitOpError("require tile sizes to be specified");
+    return DiagnosedSilenceableFailure(
+        reportUnknownTransformError(exportOp.value()));
+  }
+
+  IRRewriter rewriter(getContext());
+  if (failed(lowerWorkgroupCountComputingRegion(rewriter, exportOp.value(),
+                                                mixedTileSizes))) {
+    exportOp.value()->emitOpError("failed to lower workgroup count region");
+    return DiagnosedSilenceableFailure(
+        reportUnknownTransformError(exportOp.value()));
+  }
+
+  /// Everything from here is copied over from TileToForeachThreadOp::apply
```

Comment:
Please factor this functionality out in core and reuse, it is subject to constant evolution, not looking deeper here for now.

Reply:
I tried to, but the code here is using several methods that live on the op instance. It seemed pretty hard to decouple. If possible, I'd land this as is, and let folks who understand this better try to unify it?

Reply:
Done now.